### ISSUE TITLE
reconcile profiling/garbage-collection arguments and build options

### ISSUE THREAD

**Issue by joemfb** [urbit/urbit#1030]
As discussed in #1029, it's a little awkward to have preprocessor defines / build options and then separately have runtime CLI arguments for profiling and garbage collection. Would it be fine to remove the arguments and always profile/gc in the presence of the build options?

/cc @cgyarvin

**Comment by ohAitch** [comment]
It seems significantly preferable to only profile in the presence of
arguments, and always have the build options enabled, from a UI perspective

That said, presumably the build options are there because there's
significant associated runtime cost, even if not the whole cost of actually
profiling. I think for -g there is a whenever-invokable memory overhead,
and a large per-invocation pier exit time cost; and for -P there is a small
whenever-invokable runtime overhead, and a larger when-invoked runtime
overhead(and terminal spew on pier exit time cost)?

**Comment by belisarius222** [comment]
Yes, U3_MEMORY_DEBUG changes the shape of memory allocations, so it won't
load a loom that was created without it.

And yes, as @cgyarvin said, there is some runtime cost in checking for
U3_CPU_DEBUG all the time when it's off, so it's best left as a build
option so those checks can be preprocessed out of the final binary.

Personally I've run into more problems from forgetting to turn on the -P
flag than I have from running profiling on a build when I didn't need it. I
wouldn't mind having profiling always on when compiled that way. You can
always run with a non-profiling binary for a while if you want to turn off
CPU profiling temporarily. It's only the memory profiling where you need a
new image.


—
~rovnys-ricfer
https://urbit.org

**Comment by ohAitch** [comment]
Oh, I suppose keeping a separate urbit-P executable around is one way to do
flags, sure

**Comment by zalberico** [comment]
I think there's likely something here that could be distilled into an actionable issue, but as is it's not clear what that should be. @belisarius222 @joemfb thoughts on what to do with this? Is there something I should write up for this or is this an old conversation I should just close?

**Comment by joemfb** [comment]
This issue is a placeholder for that distillation, which will likely be the bulk of the work at which it is gesturing.
